feat(platform-browser): Add IsolatedShadowDom encapsulation

ryan-bendel opened the pull request:
Add IsolatedShadowDom encapsulation mode, which fixes style leakage by removing sharedstyleshost from dom-renderer IsolatedShadowDom class. 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

IsolatedShadowDom follows the spec, shadowdom does not now add `<style>` tags unrelated/not imported/defined by the component, which is caused by the `sharedstyleshost`. 

Updated docs.

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: [57801](https://github.com/angular/angular/issues/57801)

Follow up to the PR discussion in https://github.com/angular/angular/pull/62357

The current Shadowdom implementation is incorrect, and does not adhere to the spec. Shadowdom components should be encapsulated with no style leakage, but currently, external styles are added that are completely unrelated to the shadow component. This causes significant bloat.

### Before:
<img width="385" alt="Screenshot 2025-06-27 at 15 32 02" src="https://github.com/user-attachments/assets/0b5e2f43-f9dc-4bd3-951c-a0ef8c3fe374" />

### After:
<img width="317" alt="Screenshot 2025-06-28 at 07 12 54" src="https://github.com/user-attachments/assets/380de607-1b3c-4cf6-97a2-7c3ff9735bb1" />

## What is the new behavior?

To ensure this is a smooth transition for users, the old behaviour has been preserved in a the ShadowDom encapsulation class. `IsolatedShadowDom` has been created to allow users to implement the fixed shadowdom behaviour, which does not use sharedstyleshost.

ryan-bendel commented:
> As is, this is considered a breaking change as the `ShadowDom` encapsulation changes behavior (even with the migration).
> 
> This probably needs to be discussed if the best is to keep ShadowDom as is (would reduce the burden of landing the change) and introduce a new symbol for the new behavior.

I'll mark as breaking 👍 

Yeah the other option i mentioned was just creating a new encapsulation mode with the "fixed" shadowdom

But I leaned towards this as any new apps/components spun up would get the fixed behaviour, and it was suggested in the other PR it would be a lot of work for the team/users to have the change everything if you landed _just_ the fixed version. By marking it as deprecated users would at least be nudged in the right direction, and hopefully make the Angular teams job a bit easier in the future 🤞

dgp1130 commented:
Regarding priority of this PR: I was recently exploring using shadow DOM to isolate styles across multiple applications within the same page (MicA). One challenge is that `CdkOverlay` renders outside the main application root, so any styles are lost there. I had explored creating a shadow root there as well and noticed the same problem that styles for the entire application are unnecessarily duplicated in the extra root. I think this PR might be useful for that effort, but I'm not yet sure if using shadow roots is the right approach for the broader style isolation problem I'm looking at in that context, hopefully I can get more insight on that soon.

Regarding this PR more specifically: I wonder if this setting should be configured via dependency injection rather than creating a new `ViewEncapsulation` mode? It feels easier to configure a single token rather than migrating each component individually.

A token does potentially make it a bit harder to incrementally roll out this change, but I suspect it's not _that_ breaking and maybe doesn't _need_ to have a super fine-grained opt-out? You could certainly use DI to opt-out specific components, though the hierarchy is kind of working against you there since opting-out one component naturally opts-out its dependencies, which isn't really what you want.

I'm mainly just asking whether this is worth a new `ViewEncapsulation` mode or whether we can do something simpler with DI? I'll defer to other framework folks on whether that's worth seriously considering, just an idea.

thePunderWoman commented:
@ryan-bendel Looks like this needs a few goldens updated. `yarn run symbol-extractor:update` should update the bundle symbols and `yarn bazel run //packages/core:core_api.accept` should update the core symbols. Can you do that and then re-push? That should get CI closer to green.

ryan-bendel commented:
I'll look at the failing tests, will run the same command locally the action is using.

Not 100% clear on what is needed in some targets, for example, `pkg_npm` only has one migration (and my new one) defined, but the migrations folder has multiple migrations in it. So likely i'm misunderstanding what is supposed to be defined where.

thePunderWoman commented:
@ryan-bendel Looks like there were some recent change with these bazel rules. `interop_deps` can be rolled into `deps`. I updated it and pushed, and I also collapsed the commits into the one. Hopefully it should not have issues with that during build anymore.

ryan-bendel commented:
> @ryan-bendel Looks like there were some recent change with these bazel rules. `interop_deps` can be rolled into `deps`. I updated it and pushed, and I also collapsed the commits into the one. Hopefully it should not have issues with that during build anymore.

@thePunderWoman you're a star ⭐ - I _think_ i tried that in a previous commit, and may have ended up failing elsewhere so i reverted, lets see how it goes this time 🤞

thePunderWoman commented:
@ryan-bendel looks like it did get past that problem, but now has ✨other✨ problems.

thePunderWoman commented:
@ryan-bendel it shouldn't matter, but definitely squash your commits so we have a clean history for this.

ryan-bendel commented:
@thePunderWoman 👍 done 😃

ryan-bendel committed:
feat(platform-browser): Add IsolatedShadowDom encapsulation method

IsolatedShadowDom encapsulation fixes style leakage in Shadowdom encapsulation by removing sharedstyleshost from dom-renderer IsolatedShadowdom class. Updates docs.

thePunderWoman commented:
@ryan-bendel Ah looks like you'll need to fetch and rebase.

ryan-bendel commented:
@thePunderWoman - that, and trailing comma's, the bane of my life 😆

thePunderWoman commented:
> Hmmm `//adev/shared-docs/pipeline/guides/testing/mermaid:unit_tests` pass locally 🤔 - not sure if this a me thing

It looks like a flake. I've triggered a re-run.

ryan-bendel commented:
@thePunderWoman - i've made those changes, updated relevant docs and the PR description 👍

JeanMeche commented on adev/src/content/cli/help/generate.json:
```diff
@@ -372,7 +372,8 @@
             "Emulated",
             "None",
-            "ShadowDom"
+            "ShadowDom",
+            "IsolatedShadowDom"
           ],
           "description": "Sets the view encapsulation mode for the component. This determines how the component's styles are scoped and applied."
         }
```
Those files are updated automatically, this will require a change on the angular-cli repo.

ryan-bendel replied:
Ah my bad, removed that

JeanMeche commented on adev/src/content/cli/help/new.json:
```diff
@@ -184,7 +184,8 @@
-        "ShadowDom"
+        "ShadowDom",
+        "IsolatedShadowDom"
       ],
```
Same as above

JeanMeche commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -793,7 +793,8 @@ function conditionallyAnnotateNodePath(
   const instance = lView[CONTEXT];
   return instance?.constructor
-    ? getComponentDef(instance.constructor)?.encapsulation === ViewEncapsulation.ShadowDom
+    ? getComponentDef(instance.constructor)?.encapsulation === ViewEncapsulation.ShadowDom ||
+        getComponentDef(instance.constructor)?.encapsulation === ViewEncapsulation.IsolatedShadowDom
```
We can probably refactor that to only invoke `getComponentDef` once.

ryan-bendel commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -583,6 +597,61 @@ class ShadowDomRenderer extends DefaultDomRenderer2 {
+  override parentNode(node: any): any {
+    return this.nodeOrShadowRoot(super.parentNode(this.nodeOrShadowRoot(node)));
+  }
+}
```
Yep, that's absolutely right. That was the thing sticking all the extra styling in.

dgp1130 commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -583,6 +597,61 @@ class ShadowDomRenderer extends DefaultDomRenderer2 {
+
+      styleEl.textContent = style;
+      this.shadowRoot.appendChild(styleEl);
+    }
```
My guess is that this would be for JIT rendering `styleUrl`, but I'm not 100% on that. I don't suppose @thePunderWoman would have more context on when `component.getExternalStyles` returns content?

Regardless I'm inclined to keep `ShadowDom` and `IsolatedShadowDom` as aligned as possible, unless we have reason to believe this snippet is harmful.

dgp1130 reviewed (approved):
No concerns on my end, thanks for the contribution here!

thePunderWoman reviewed (approved):
LGTM! Super huge thanks, @ryan-bendel, for your effort here! 🎉 

reviewed-for: fw-compiler, fw-general, angular-dev, public-api

thePunderWoman commented:
@ryan-bendel you have a tiny docs conflict. Once you resolve that, this should be merge ready!!!

JeanMeche reviewed (approved):
reviewed-for: public-api

robertIsaac commented:
oh this is a great feature that we desperately need

JeanMeche commented:
Fwiw we reverted that feature for 20.2 and we're looking for shipping it in 21.0
